text/0121-data-types.md

```diff
@@ -0,0 +1,435 @@
+
+For Latin-based languages, the words are separated by space.
+For Kanji characters, the words are separated by character.
+
+For Latin-based languages, there are two kinds of space separators:
+- **Soft spaces (distance: 1)**: whitespaces, quotes, `-` | `_` | `\` | `:` | `/` | `\\` | `@` | `"` | `+` | `~` | `=` | `^` | `*` | `#`
+- **Hard spaces (distance: 8)**: `.` | `;` | `,` | `!` | `?` | `(` | `)` | `[` | `]` | `{` | `}`| `|`
+
+Distance plays an essential role in determining whether documents are relevant. The `proximity` ranking rule sorts the results by increasing distance between matched query terms. Two words separated by a soft space are closer and thus considered more relevant than two words separated by a hard space.
+
+After the tokenizing process, each word is indexed and stored in the global dictionary of the corresponding index.
+
+#### 3.1.1. Examples
+
+To demonstrate how a string is split by space, let's say you have the following string as an input:
+
+`"Bruce Willis,Vin Diesel"`
+
+In the example above, the distance between `Bruce` and `Willis` is equal to `1`. The distance between `Vin` and `Diesel` is equal to `1` too.
+But, the distance between `Bruce` and `Vin` is equal to `8`. The same goes for `Bruce` and `Diesel`, or `Willis` and `Vin`, or also `Willis` and `Diesel`.
+
+Let's see another example. Given two documents:
+
+```json
+[
+    {
+        "movie_id": "001",
+        "description": "Bruce.Willis"
+    },
+    {
+        "movie_id": "002",
+        "description": "Bruce super Willis"
+    }
+]
+```
+
+When making a query on `Bruce Willis`, `002` will be the first document returned and `001` will be the second one.
+This will happen because the proximity distance between `Bruce` and `Willis` is equal to `2` in the document `002` whereas the distance between `Bruce` and `Willis` is equal to `8` in the document `001` since the full stop is a hard space.
+
+### 3.2. Array Tokenization
+
+An array is recursively broken into separate string tokens, which means separate words. After the tokenizing process, each word is indexed and stored in the global dictionary of the corresponding index.
+
+#### 3.2.1. Examples
+
+The following input:
+
+```json
+[
+    [
+        "Bruce Willis",
+        "Vin Diesel"
+    ],
+    "Kung Fu Panda"
+]
+```
+
+Will be processed as if all elements were arranged at the same level:
+
+```json
+"Bruce Willis. Vin Diesel. Kung Fu Panda."
+```
+
+The strings above will be separated by soft and hard spaces exactly as explained in the string example.
+
+> See [3.1.1. Examples section](#311-examples).
+
+### 3.3. Object
+
+#### 3.3.1. Example
+
+The following input:
+
+```json
+{
+    "movie_id": "1564saqw12ss",
+    "title": "Kung fu Panda"
+}
+```
+
+In the example above, `movie_id`, `1564saqw12ss`, `title`, `Kung fu Panda` are all considered sentences. The colon `:` and comma `,` characters are used as separators.
+
+`"movie_id. 1564saqw12ss. title. Kung fu Panda."`
```

Comment:
In the paragraph above we say: 

> The colon `:` and comma `,` characters are used as separators.

With that in mind, shouldn't this code sample be:

```suggestion
`"movie_id: 1564saqw12ss. title: Kung fu Panda."`
```

Reply:
Indeed @guimachiavelli, it is not clear.

What it says here is that the keys and values of the document are written this way internally. The `:`, `,` of the JSON document are not kept and the different keys and values are separated by a `.`.

I will modify this section.

Reply:
Ah, I see! Your latest edit makes it much clearer, thanks!

Reply:
@Kerollmops Can you confirm that objects are represented this way internally?

Reply:
@guimachiavelli The spec has been updated!